### PR TITLE
fixed limit for publish frequency

### LINKED ISSUE #372: s300 publish rate throttling does not work

### PR DESCRIPTION
**Pull request by kalectro**
- fixes #372 
 - new source code files added (proper [APACHE license header](https://github.com/ipa320/setup/tree/master/templates) **must** be used)

### CODE REVIEW

**Comment by ipa-rmb** [comment]
Many thanks for finding this bug. I was so free to append a change on the variable name loop_rate_, which was no loop rate, to make it more meaningful for the given purpose.

**Review by ipa-rmb — DISMISSED** [review]

File: cob_sick_s300/ros/src/cob_sick_s300.cpp
Comment (fmessmer) [inline cob_sick_s300/ros/src/cob_sick_s300.cpp:223]: I think we should set a variable
```
ros::Time now = ros::Time::now();
```
right at the beginning of the publish function  and then use `now` throughout the function instead of getting new times everytime

**Review by fmessmer — CHANGES_REQUESTED** [review]

**Review by ipa-rmb — APPROVED** [review]

**Comment by fmessmer** [comment]
@kalectro do you still see yourself in this PR? :wink:

**Comment by ipa-rmb** [comment]
@ipa-fxm are you fine with the PR as well? Please approve then. I will merge after @kalectro and you confirmed.

**Comment by ipa-rmb** [comment]
@fmessmer if you do not have any further objections, I will merge it tomorrow.

**Comment by fmessmer** [comment]
@ipa-rmb @kalectro 
I have a bunch of PRs to test on HW next week...
thus, let me test this one, too...I'll merge after testing

**Review by fmessmer — CHANGES_REQUESTED** [review]
I tested this PR today...it seems there is a need for a little more discussion:

on our robot we have `publish_frequency` set to `12Hz` and - with current `ipa320/indigo_dev` - we receive a topic rate of `25Hz` (using `rostopic hz /base_laser_front/scan_raw`)
with this PR by @kalectro, I get a topic rate of only `9.7Hz` (using `rostopic hz /base_laser_front/scan_raw`)

thus, I see that it hasn't been correct before (i.e. "doubled" `publish_frequency`), but it's not correct now either (i.e. `9.7Hz` != `12Hz`)

also, the "doubled" `publish_frequency` has been around for quite a while and I think our applications, e.g. navigation, rely on the topic rate of `25Hz`

however, setting the `publish_frequency` to `25Hz`, with this PR I only get a topic rate of `13.5Hz`
while with `ipa320/indigo_dev` the topic rate is `25Hz` again (wtf?)...it's even `25Hz` with `publish_frequency` set to `50Hz`...and it's `25Hz` with `publish_frequency` set to `5Hz`
thus, it seems that in current `ipa320/indigo_dev` the `publish_frequency` parameter has no actual influence on the publish rate at all...it's `25Hz` for any value!

as all this seems rather confusing, I would like to here some other maintainer's opinion (@ipa-fez @ipa-jba @floweisshardt @benmaidel) before merging!

BTW, what's the default/max publish_rate of the actual HW? is it configurable at all?

@mgruhler as you have been working with the sensor quite a bit, do you have anything to add to the discussion?

**Comment by ipa-fez** [comment]
From a quick glance it looks like we just skip scans if the HW sends scans too fast, so if the hardware runs at 25 Hz and we request 20 Hz, we'll skip basically every other scan because the 1/20 ms haven't passed yet.

Maybe this should always publish at HW speed and topic_tools throttle can be used if lower HZ is required.

**Comment by mgruhler** [comment]
I never looked into this parameter.

But I can share some insights (that are almost not documented in the Sick docus) about the possible frequencies and what influences those that I once added in https://github.com/ipa320/cob_driver/issues/338#issuecomment-311023596.

To summarize:
- Max Frequency is 25Hz for new protocol and 12.5Hz for old protocol scanners
- There are many settings in the Sick configuration software that further limits this

Also, when looking at the code, I guess that the parameter is effectively a noop, the way it is implemented right now:

```C++
if(ros::Time::now()-loop_rate_.now()>=ros::Duration(1./publish_frequency))
```
is actually calling `now()` on the `ros::Time loop_rate_` (which is probably the most misleading variable name i've ever encountered) object. I'm guessing this is returning the same as `ros::Time::now()`. Thus, this check is always false, leading to it not being used at all.
If it gets fixed though, @ipa-fez is right, we will just skip every other scan. And if you set it to something of which 25 is not a multiple of (i.e. 25/x, where x is an integer, basically everything except 12.5, 8,33333, 6,25, 5, ...), you will get scans with different spacings that could result in those strange frequencies that you see (as `rostopic hz` is actually averaging, if I'm correct).

I agree with @ipa-fez that the best way would be to remove this parameter completely and then throttle the topic if you actually need it at lower frequency.

**Comment by fmessmer** [comment]
I'm going to close this unmerged given the discussion and with #377 being merged

**Comment by kalectro** [comment]
Sorry for not replying, just saw everything today and like the solution

**Pull request closed without merging**
